File: aws-typescript/3_component-resources/solutions/exercise-2/webserver-component.ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import { ComponentResource, ComponentResourceOptions, Input, Output } from "@pulumi/pulumi";

export interface WebserverArgs {
    vpcId: Input<string>;
    subnetId: Input<string>;
    size: string;
    message: string;
};

export class Webserver extends ComponentResource {
    public readonly serverIp: Output<string>;

    // Must have a constructor that defines the parameters and namespace - "custom:x:Webserver" in this case.
    constructor(name: string, args: WebserverArgs, opts?: ComponentResourceOptions) {
        super("custom:x:Webserver", name, args, opts);

        /////
        // Declare the children resources (i.e. security group, instance) be sure to include a resource option of "parent:this".
        /////

        // Create an EC2 instance running a simple web server.
        // Get the id for the latest Amazon Linux AMI
        const ami = aws.ec2.getAmi({
            filters: [
                { name: "name", values: ["amzn-ami-hvm-*-x86_64-ebs"] },
            ],
            owners: ["137112412989"], // Amazon
            mostRecent: true,
        }).then(result => result.id);
        
        // create a new security group for port 80
        const group = new aws.ec2.SecurityGroup(`${name}-secgrp`, {
            vpcId: args.vpcId,
            ingress: [
                { protocol: "tcp", fromPort: 80, toPort: 80, cidrBlocks: ["0.0.0.0/0"] },
            ],
        }, {parent: this,
            aliases: [{ parent: pulumi.rootStackResource }]
        });
        
        // create a simple web server using the startup script for the instance
        const userData =
        `#!/bin/bash
        echo "${args.message}" > index.html
        nohup python -m SimpleHTTPServer 80 &`;
        
        let instanceType: aws.ec2.InstanceType = aws.ec2.InstanceType.T3_Micro;
        if (args.size === "large") {
            instanceType = aws.ec2.InstanceType.T3_Medium;
        } else if (args.size === "medium") {
            instanceType = aws.ec2.InstanceType.T3_Small
        }
        
        const serverName = `${name}-web-server`;
        const server = new aws.ec2.Instance(serverName, {
            subnetId: args.subnetId,
            tags: { "Name": serverName },
            instanceType: instanceType,
            vpcSecurityGroupIds: [ group.id ], // reference the group object above
            ami: ami,
            userData: userData,              // start a simple web server
            associatePublicIpAddress: true,
        }, {parent: this,
            aliases: [{ parent: pulumi.rootStackResource }]
        });

        this.serverIp = server.publicIp;

        // End with this. It is used for display purposes.
        this.registerOutputs();
    };
};

File: aws-python/3_component-resources/solutions/exercise-1/webserver_component.py
# A component resource module shell 
# See comments below for help

import pulumi_aws as aws

from pulumi import Alias, ComponentResource, ResourceOptions, Input, ROOT_STACK_RESOURCE 


class WebserverArgs:
    def __init__(self,
                vpc_id: Input[str],
                subnet_id: Input[str],
                size: str,
                message: str
                ):

        # Set the class args
        self.vpc_id = vpc_id
        self.subnet_id = subnet_id
        self.size = size
        self.message = message


class Webserver(ComponentResource):
    def __init__(self,
                 name: str,
                 args: WebserverArgs,
                 opts: ResourceOptions = None):

        # Leave this line. You can modify 'custom:resource:Webserver' if you want
        super().__init__('custom:resource:Webserver', name, {}, opts)

        if args.size == "medium":
            instance_type = aws.ec2.InstanceType.T3_MEDIUM
        elif args.size == "large":
            instance_type = aws.ec2.InstanceType.T3_LARGE
        else:
            instance_type = aws.ec2.InstanceType.T3_MICRO 

        ami = aws.ec2.get_ami(most_recent=True,
                  owners=["137112412989"],
                  filters=[aws.GetAmiFilterArgs(name="name", values=["amzn-ami-hvm-*"])])

        group = aws.ec2.SecurityGroup('web-secgrp',
            description='Enable HTTP access',
            vpc_id=args.vpc_id,
            ingress=[aws.ec2.SecurityGroupIngressArgs(
                protocol='tcp',
                from_port=80,
                to_port=80,
                cidr_blocks=['0.0.0.0/0'],
            )],
            opts=ResourceOptions(
                parent=self,
                aliases=[Alias(parent=ROOT_STACK_RESOURCE)]
            ))

        user_data = f"""
        #!/bin/bash
        echo "{args.message}" > index.html
        nohup python -m SimpleHTTPServer 80 &
        """

        server_name = f"{name}-web-server"
        server = aws.ec2.Instance(server_name,
            subnet_id=args.subnet_id,
            tags={"Name":server_name},
            instance_type=instance_type,
            vpc_security_group_ids=[group.id],
            user_data=user_data,
            associate_public_ip_address=True,
            ami=ami.id,
            opts=ResourceOptions(
                parent=self,
                aliases=[Alias(parent=ROOT_STACK_RESOURCE)]
            ))

        self.server_ip = server.public_ip
        # End with this. It is used for display purposes.
        self.register_outputs({})
